webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/NodeInfoService.scala

```diff
@@ -229,6 +236,242 @@ final case class LocalNodeInfoCache(
+              )
+              result.updated.addOne(LDAPNodeInfo(nodeEntry, nodeInv, machineInv))
+        }
+      }
+    }
+    val nbNodeEntries = result.updated.size
+    logEffect.trace(s"nodeEntries are nodeEntries: ${result.updated.mkString(",")}")
+
+    /*
+     * Loop 2: for node inventories.
+     * Only process node inventories that are not yet marked as done.
+     * Again, we can have error: here, we know the node info for these inventories must
+     * be in cache, since the one form updates were done in loop 1. If not in cache, it's
+     * an error.
+     */
+    infoMaps.nodeInventories.foreach { case (id, nodeInv) =>
+      if(!managedNodeInventories.contains(id)) {
+        currentCache.nodeInfos.get(NodeId(id)) match {
+          case None => // oups
+            result.nodeErrors.addOne(id)
+          case Some((ldapInfo, nodeInfo)) =>
+            val nodeEntry = ldapInfo.nodeEntry
+            val machineInv = getMachineInventory(nodeInv(A_CONTAINER_DN), infoMaps.machineInventories, managedMachineInventories, () => ldapInfo.machineEntry)
+            result.updated.addOne(LDAPNodeInfo(nodeEntry, nodeInv, machineInv))
+        }
+      }
+    }
+
+    val nbNodeInvs = result.updated.size - nbNodeEntries
+    logEffect.trace(s"inventoryEntries are inventoryEntries: ${result.updated.iterator.drop(nbNodeEntries).mkString(",")}")
+
+    /* loop3: for machine inventories
+     * Very similar than node inventories, for same reasons, but one machine can be mapped to several nodes!
+     */
+    infoMaps.machineInventories.foreach { case (containerDn, machineInv) =>
+      if(!managedMachineInventories.contains(containerDn)) {
+        // the tricky part: there may be several nodes with the same containerDn
+        // now, we must have AT LEAST one value in res, else it's the same kind of error than a node inventory without a node.
+        var atLeastOne = false
+        currentCache.nodeInfos.collect { case (nodeId, (ldapInfo, _)) if(ldapInfo.nodeInventoryEntry(A_CONTAINER_DN).equals(Some(containerDn))) =>
+          atLeastOne = true
+          result.updated.addOne(LDAPNodeInfo(ldapInfo.nodeEntry, ldapInfo.nodeInventoryEntry, Some(machineInv)))
+        }
+        if(!atLeastOne) {
+          result.containerErrors.addOne(containerDn)
+        }
+      }
+    }
+
+    logEffect.trace(s"machineInventoriesEntries are machineInventoriesEntries: ${result.updated.iterator.drop(nbNodeEntries+nbNodeInvs).mkString(",")}")
+
+    result
+  }
+
+  /*
+   * Simpler version of `constructNodesFromPartialUpdate` where we assume that we have all datas,
+   * so we can only loop one time for nodes and be done.
+   */
+  def constructNodesFromAllEntries(infoMaps: InfoMaps, checkRoot: Boolean = true): IOResult[NodeUpdates] = {
+    for {
+      results      <- Ref.make(NodeUpdates())
+      rootMissing  <- Ref.make(checkRoot)
+      _            <- ZIO.foreach(infoMaps.nodes) { case (id, nodeEntry) =>
```

Comment:
what happens if a nodes is not yet totally inserted at creation, and we have  its container, but nothing else? Won't it silenty fails ?

Reply:
ha it would, but it would catch up when the node is inserted, or fail there and catchup at the next time